datapath: Create sysctl `rp_filter` overwrite config on agent init

dylandreimerink opened the pull request:
SystemD versions greater than 245 will create sysctl config which sets
the `rp_filter` value for all network interfaces to 1. This conflicts
with cilium which requires `rp_filter` to be 0 on interfaces it uses.

This commit adds a small utility/tool: `sysctlfix` which will insert
a config file into the `/etc/sysctl.d` dir with the highest priority
containing directives to disable `rp_filter` and perhaps to contain
other sysctl config in future.

This utility is called as an init container before the cilium agent
starts. Because the sysctl config is in place before the agent starts,
all interfaces created by the agent and matching the patten in the
config file will have `rp_filter` disabled, even when SystemD >=245 is
installed.

Fixes: #10645
Fixes: #19909

```release-note
Fixed SystemD >=245 sysctl(`rp_filter`) config incompatibility
```

dylandreimerink committed:
datapath: Create sysctl `rp_filter` overwrite config on agent init

SystemD versions greater than 245 will create sysctl config which sets
the `rp_filter` value for all network interfaces to 1. This conflicts
with cilium which requires `rp_filter` to be 0 on interfaces it uses.

This commit adds a small utility/tool: `sysctlfix` which will insert
a config file into the `/etc/sysctl.d` dir with the highest priority
containing directives to disable `rp_filter` and perhaps to contain
other sysctl config in future.

This utility is called as an init container before the cilium agent
starts. Because the sysctl config is in place before the agent starts,
all interfaces created by the agent and matching the patten in the
config file will have `rp_filter` disabled, even when SystemD >=245 is
installed.

Fixes: #10645
Fixes: #19909
Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

aanm commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,59 @@
+// we can overwrite any other config and ensure correct sysctl options for Cilium to function.
+
+const (
+	sysctlD = "/etc/sysctl.d/"
```
Is this directory always the same? Should we make it user-configurable by passing it as a flag?

dylandreimerink replied:
There are multiple paths defined which overwrite each other in a set order. So as far as I can tell these directories are always the same and the `/etc/sysctl.d` overwrites all others.

From https://man7.org/linux/man-pages/man5/sysctl.d.5.html
```
       Configuration files are read from directories in /etc/, /run/,
       /usr/local/lib/, and /usr/lib/, in order of precedence, as listed
       in the SYNOPSIS section above. Files must have the ".conf"
       extension. Files in /etc/ override files with the same name in
       /run/, /usr/local/lib/, and /usr/lib/. Files in /run/ override
       files with the same name under /usr/.

       All configuration files are sorted by their filename in
       lexicographic order, regardless of which of the directories they
       reside in. If multiple files specify the same option, the entry
       in the file with the lexicographically latest name will take
       precedence. Thus, the configuration in a certain file may either
       be replaced completely (by placing a file with the same name in a
       directory with higher priority), or individual settings might be
       changed (by specifying additional settings in a file with a
       different name that is ordered later).
```

dylandreimerink replied:
I made flags for all of the constants anyway, none of them are required, so we use default values if no flags are passed. It doesn't hurt and might be useful to have it in the binary in case we ever need it in a production environment for some reason.

sayboras reviewed (approved):
LGTM :100:

aditighag reviewed (approved):
🚀

aditighag commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,59 @@
+
+var sysctlConfig = strings.Join([]string{
+	"# Disable rp_filter on Cilium interfaces since it may cause mangled packets to be dropped",
+	"net.ipv4.conf.lxc*.rp_filter = 0",
```
I overlooked that we are completely disabling the filter. What about setting it to the `Loose` mode? That's all we should need to pass the packets with asymmetric routing. Details - https://github.com/cilium/cilium/issues/13130. That's the value we are currently setting as well.

dylandreimerink replied:
I set it to disabled since that seemed to be the suggested fix in #10645. According too @aanm, [this](https://github.com/systemd/systemd/commit/230450d4e4f1f5fc9fa4295ed9185eea5b6ea16e) commit in SystemD is the source of the issue. In the commit the `rp_filter` is changed from 1(Strict) to 2(Loose). In #10645 [this](https://github.com/systemd/systemd/commit/5d4fc0e665a3639f92ac880896c56f9533441307#diff-7816eed8ca6324f23a690cc5f58e6bf7) commit is mentioned as the culprit (also a change to 2(Loose))

The description of 'Loose' mode is:
> Loose mode as defined in RFC3704 Loose Reverse Path
        Each incoming packet's source address is also tested against the FIB
        and if the source address is not reachable via any interface
        the packet check will fail.
        
I am not sure if there are common use cases in Cilium where we handle packets for source addresses which don't appear in the FIB, but it wouldn't surprise me if there are.

aditighag replied:
Cilium isn't the only entity that could be running on a host. The issues you've linked above indicate that systemd overwrites `rp_filter` settings set by Cilium. Regardless, I would like for us to understand why we would need to disable source address checks altogether in case of Cilium.

dylandreimerink committed:
datapath: Reload `systemd-sysctl` after changing config

This commit adds logic which triggers the  `systemd-sysctl` unit to
reload the config and apply it. This is necessary for any existing
network interfaces. To avoid runtime dependencies on the
`systemctl` command we trigger the reload of the unit directly via
the D-bus.

The tool is now idempotent, it only updates the config file it doesn't
match the desired state, and only triggers a systemd unit reload if the
config changed.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

brb commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,142 @@
+
+var sysctlConfig = strings.Join([]string{
+	"# Disable rp_filter on Cilium interfaces since it may cause mangled packets to be dropped",
+	"net.ipv4.conf.all.rp_filter = 0",
```
Aren't we changing the rp_filter setting of other non-Cilium ifaces? If yes, then this might be undesirable behavior by users who are concerned about the packet spoofing.

dylandreimerink replied:
Yes, it does. According to @rastislavs this is required to fix #19761. Perhaps we can find a middle ground between disabling on all interfaces and just the Cilium interfaces? 

If `rp_filter` on host interfaces really does create connectivity problems, perhaps we should disable it and fast-track spoofing detection in Cilium, so we can guarantee it is compatible?

brb replied:
> According to @rastislavs this is required to fix

I think it makes sense to first understand which packet path was broken due to rp_filter=2 in that case.

dylandreimerink replied:
That seems fair, I will revert this PR to only setting `rp_filter` for Cilium interfaces so we can get this PR merged for the other issues. Then we will do a followup PR once we have more details regarding #19761.

rastislavs replied:
But isn't Cilium already setting `net.ipv4.conf.all.rp_filter` to `0` upon its startup ([here](https://github.com/cilium/cilium/blob/v1.11.5/pkg/datapath/loader/base.go#L245))? At least this is a log message that I can see in Cilium logs when it starts:
```
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.all.rp_filter sysParamValue=0
```
So why should this be any different to that logic?

aditighag replied:
We set the config to the `Loose` mode while installing ENI rules on EKS clusters - http://github.com/cilium/cilium/blob/master/pkg/datapath/loader/base.go#L146-L146. We should ideally have a consistent way of setting the config. I would suggest investigating this further. You may refer to this issue https://github.com/cilium/cilium/issues/13130 to get pointers for debugging.

dylandreimerink replied:
> But isn't Cilium already setting net.ipv4.conf.all.rp_filter to 0 upon its startup

Yes, but this only changes the runtime option, if systemd is triggered for any reason it can just change the runtime setting to match the config. By setting the desired sysctl parameter in systemd config, we ensure that the value stays even after systemd config reloads

Edit: nevermind, you are hinting at the part where we set the setting for all interfaces, disregard my comment above.

rastislavs replied:
Please see my latest comment on https://github.com/cilium/cilium/issues/19761#issuecomment-1150751732

sayboras replied:
https://github.com/cilium/cilium/pull/20072#issuecomment-1156495539

brb reviewed (approved):
Reviewed only two lines (`...lxc*.rp_filter = 0` and `...cilium_*.rp_filter = 0`), LGTM!

tklauser commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,139 @@
+	)
+)
+
+var sysctlConfig = strings.Join([]string{
```
Nit: given this is a constant string, I think this could be defined as a const multi-line string to avoid `strings.Join` altogether.

dylandreimerink replied:
This style initially made sense to interleave go comments and configuration. But that is not required due to recent changes, and we can put any comments in the file itself anyway like done new. So agreed!

tklauser commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	// Move offset into the file to the beginning
+	_, err = f.Seek(0, 0)
```
Let's use a named constant for `whence` which should also make the comment superfluous:

```suggestion
	_, err = f.Seek(0, io.SeekStart)
```

dylandreimerink replied:
That is neat, didn't know about those constants

tklauser commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,139 @@
+
+	fmt.Println("sysctl config created/updated")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```
1 minute seems like a rather long timeout. How was this determined?

dylandreimerink replied:
You are probably right. Didn't put much thought into the actual value, 1 minutes does seem like a bit on the long side for a startup program. I will change it to 15 secondes or so. That feels like a more reasonable for timeout SystemD to reload the unit. In my experience this is normally a sub-second thing.

tklauser reviewed (commented):
Some non-blocking nits inline.

tklauser reviewed (approved):
Thanks

christarazi commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -454,6 +455,49 @@ spec:
+        # same directory where we install cilium cni plugin so that exec permissions
+        # are available.
+        - |
+          cp /usr/bin/cilium-sysctlfix /hostbin/cilium-sysctlfix;
+          nsenter --mount=/hostproc/1/ns/mnt "${BIN_PATH}/cilium-sysctlfix";
+          rm /hostbin/cilium-sysctlfix
```
Does there need to be a semi-colon at the end of these lines?

dylandreimerink replied:
Not sure to be honest :sweat_smile:, I copied this part from the `cilium-mount` container at line 429, I haven't tried it without to see how it gets interpreted

christarazi reviewed (commented):
Code LGTM at a glance. I didn't do a deep review. 

1. Minor nit comment below
2. Could you update the commits and PR desc. to now reference the CNI approach rather than init container? Could you also add the background around the lb-only flag and only disabling `rp_filter` for `cilium*` and `lxc*` interfaces, rather than `all`, etc?
3. This might be a dumb question, but I don't see how this is getting invoked each time the CNI plugin is invoked. Could you point it out?

aditighag commented on tools/sysctlfix/main.go:
```diff
@@ -0,0 +1,136 @@
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+
+	conn, err := dbus.NewSystemdConnectionContext(ctx)
```
Ah! Nice to see the API being exposed via dbus.

aditighag reviewed (commented):
Mainly looked at changes in `tools/sysctlfix/main.go`. LGTM!
Do we still need changes related to setting `rp_filter` configs - https://github.com/cilium/cilium/blob/master/bpf/init.sh#L63?

ldelossa reviewed (approved):
LGTM

sayboras reviewed (approved):
LGTM :100:
